### PR TITLE
Added :int-mod and :int-dec instructions, and a few more tests.

### PR DESCRIPTION
**Pull request by thelmuth**
I found these instructions from a year ago, got them in the new format, and added some tests. Helped me get my feet wet with the new testing, and didn't want others to replicate this work. Should be an easy merge.

### CODE REVIEW

File: src/puj/push/instructions/numeric.clj [hunk 1]
```diff
@@ -21,8 +21,18 @@
      (i/simple-instruction :int-mult [:int :int] :int 0 *)
      {:puj.push.instruction/doc  "Pushes the product of the top 2 ints onto the int stack."})
 
+   ; @TODO :int-div - what kind of protected division should we use? push-like, which would no-op if divide by zero, or analytic quotient, as described here: https://push-language.hampshire.edu/t/gptp-2019-notes/
+
+   (with-meta
+     (i/simple-instruction :int-mod [:int :int] :int 0 mod)
```
Comment (erp12) [inline src/puj/push/instructions/numeric.clj:34]: I believe we need to use a custom function to guard against mod by zero, right?

File: test/puj/push/core_instructions_test.clj [hunk 2]
```diff
@@ -34,10 +42,30 @@
...
+      {:int '(5 -33 -12)}
+      {:int '(2 -12)}
+
       :int-inc
       {:int '(50)}
       {:int '(51)}
 
+      :int-inc      ;; Test empty stack
+      {:int '()}
+      {:int '()}
```
Comment (erp12) [inline test/puj/push/core_instructions_test.clj:63]: Not something that needs to change, but just an fyi.

Now that the code for popping arguments and pushing return values is centralized for all simple instructions, we don't need to test this edge case for every instruction. Instead we test it once here:

https://github.com/push-language/Puj/blob/daaa4adf5f6cd9535785bcc2ce970933ecd1cca0/test/puj/push/instruction_test.clj#L22

**Review by erp12 — CHANGES_REQUESTED** [review]

File: src/puj/push/instructions/numeric.clj (hunk 1 above)
Comment (thelmuth) [reply src/puj/push/instructions/numeric.clj:34]: You are correct! In Clojush, we just no-op, which I don't like for the same reasons I don't like no-op for divide-by-zero. Maybe mod by zero should give zero?

I also remembered that Clojure's `mod` sometimes gives `##INF` and `##NAN` when modding large numbers 😭 . See this: https://push-language.hampshire.edu/t/inf-or-nan-on-float-stack/1540?u=thelmuth

That means we need to to something to ensure they don't appear on the stacks. What I did for Clojush, which might not be the best solution, is here: https://github.com/lspector/Clojush/pull/292/files

All told, maybe we need to have more discussion about what we want the desired behavior to be before adding `int_mod`. I propose, for now, that I replace this with a TODO like with `int_div`, and point to this discussion so we remember when coding it.

File: test/puj/push/core_instructions_test.clj (hunk 2 above)
Comment (thelmuth) [reply test/puj/push/core_instructions_test.clj:63]: Very cool! But, I can't figure out what that line does, without more context, and would have no idea that it tests (all? some? one?) instructions on an empty stack if I encountered it in the wild. @erp12, some of the testing code could use some more comments/docstrings please 😄

File: src/puj/push/instructions/numeric.clj (hunk 1 above)
Comment (thelmuth) [reply src/puj/push/instructions/numeric.clj:34]: I pushed a commit that does this, but not sure about how to mark this as resolved.

EDIT: Oops, forgot to remove `:int-mod` test, which is failing. Will do in a second.

EDIT2: Done.

Comment (erp12) [reply src/puj/push/instructions/numeric.clj:34]: 👍

File: test/puj/push/core_instructions_test.clj (hunk 2 above)
Comment (erp12) [reply test/puj/push/core_instructions_test.clj:63]: Good point. I will work on that shortly in a separate PR.

The `instruction_test.clj` file doesn't test any of the instructions. Instead it tests the logic of the instruction abstractions (ie. `simple-instruction`). In the case of empty stacks (or insufficient args), we always expect the push-state to be unchanged and there is one code path that controls this behavior for all simple instructions. Thus, IMHO, we don't need to test any specific instructions.

This gets more complicated for more generic instruction types. For example the `state-to-state-instruction` has free range over the push-state and would call for more tests than a `simple-instruction`.

**Comment by erp12** [comment]
Looks good to me. I would recommend a "Squash and merge" for this PR so that the failing tests don't appear in the history of master.

File: test/puj/push/core_instructions_test.clj (hunk 2 above)
Comment (thelmuth) [reply test/puj/push/core_instructions_test.clj:63]: Sounds good! Is it worth testing one instruction with empty stacks, to make sure that the actual execution of the instruction doesn't differ from this abstract testing without an instruction? If not, it would be fine to remove this!
